### PR TITLE
NPM distribution

### PR DESCRIPTION
**Pull request by samelhusseini**
This PR prepares tasks that help us release our compressed built files as UMD / CommonJS modules for distribution Blockly on NPM.

Note: this PR is against master. I've had to also pull in the Typescript blockly.d.ts changes defined in https://github.com/google/blockly/pull/2738.

## The basics

- [x] I branched from ~~develop~~master
- [x] My pull request is against ~~develop~~master
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

https://github.com/google/blockly/issues/177

### Proposed Changes

Package our built files into UMD modules.

### Reason for Changes

In order to release Blockly on NPM.

### Test Coverage

I've build samples for different development environments at https://github.com/google/blockly-samples and tested locally via ``npm link ...``

### Documentation

I've added documentation throughout. 
The README.md file in this PR is what users will see that go to https://www.npmjs.com/package/blockly. 
There is a bunch more documentation at https://github.com/google/blockly-samples. 

I've also added comments in the gulpfile.js in this PR that describes what is getting built by each task and why.

### Additional Information

Yay! NPM!

### CODE REVIEW

File: package/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+## Samples
+
+For samples on how to integrate Blockly into your project, view the list of samples at [blockly-samples](https://github.com/google/blockly-samples).
+
+
+### Importing Blockly
+
+When you import Blockly with ``import * as Blockly from 'blockly';`` you'll getting the default modules: Blockly core, Blockly built-in blocks, the JavaScript generator and the English lang files.
```
Comment (RoboErikG) [inline package/README.md]: s/getting/get/

Comment (RoboErikG) [reply package/README.md]: and nit:line length

File: typings/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# blockly.d.ts
+
+This ``blockly.d.ts`` file describes the TypeScript type definitions for Blockly.
+If you consume Blockly through ``npm``, this file is already included in the ``npm`` package.
+Otherwise, you can include a copy of this file in your with your sources and reference it through a [Triple-Slash directive](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html).
```
Comment (RoboErikG) [inline typings/README.md]: s/in your with/with/

File: package/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+If you need more flexibility, you'll want to define your imports more carefully: 
+
+#### Blockly Core
+
+```js
+import * as Blockly from 'blockly/core';
+```
+
+#### Blockly build in blocks
```
Comment (rachel-fenichel) [inline package/README.md]: build -> built

**Pull request merged**